ggshield/core/git_hooks/ci/previous_commit.py

```diff
@@ -0,0 +1,141 @@
+import os
+from typing import Optional
+
+import click
+
+from ggshield.core.errors import UnexpectedError
+from ggshield.core.git_shell import get_last_commit_sha_of_branch
+from ggshield.core.utils import EMPTY_SHA
+
+from .supported_ci import SupportedCI
+
+
+def get_previous_commit_from_ci_env(
+    verbose: bool,
+) -> Optional[str]:
+    """
+    Returns the previous HEAD sha of the targeted branch.
+    Returns None if there was no commit before.
+    """
+    if os.getenv("GITLAB_CI"):
```

Comment:
nit: we are duplicating the CI detection rules here and in commit_range.py. You can factorize this as a static method in SupportedCI like this:

```diff
diff --git a/ggshield/core/git_hooks/ci/supported_ci.py b/ggshield/core/git_hooks/ci/supported_ci.py
index 647c6fd..27612c1 100644
--- a/ggshield/core/git_hooks/ci/supported_ci.py
+++ b/ggshield/core/git_hooks/ci/supported_ci.py
@@ -1,5 +1,9 @@
+import os
+
 from enum import Enum
 
+from ggshield.core.errors import UnexpectedError
+
 
 class SupportedCI(Enum):
     GITLAB = "GITLAB"
@@ -10,3 +14,26 @@ class SupportedCI(Enum):
     BITBUCKET = "BITBUCKET PIPELINES"
     DRONE = "DRONE"
     AZURE = "AZURE PIPELINES"
+
+    @staticmethod
+    def from_ci_env() -> "SupportedCI":
+        if os.getenv("GITLAB_CI"):
+            return SupportedCI.GITLAB
+        if os.getenv("GITHUB_ACTIONS"):
+            return SupportedCI.GITHUB
+        if os.getenv("TRAVIS"):
+            return SupportedCI.TRAVIS
+        if os.getenv("JENKINS_HOME") or os.getenv("JENKINS_URL"):
+            return SupportedCI.JENKINS
+        if os.getenv("CIRCLECI"):
+            return SupportedCI.CIRCLECI
+        if os.getenv("BITBUCKET_COMMIT"):
+            return SupportedCI.BITBUCKET
+        if os.getenv("DRONE"):
+            return SupportedCI.DRONE
+        if os.getenv("BUILD_BUILDID"):
+            return SupportedCI.AZURE
+        raise UnexpectedError(
+            f"Current CI is not detected or supported."
+            f" Supported CIs: {', '.join([ci.value for ci in SupportedCI])}."
+        )
```

Then this code can use a dict to dispatch to the correct function:

```python
PREVIOUS_COMMIT_SHA_FUNCTIONS = {
    SupportedCI.GITLAB: gitlab_previous_commit_sha,
    SupportedCI.GITHUB: github_previous_commit_sha,
}

def get_previous_commit_from_ci_env(
    verbose: bool,
) -> Optional[str]:
    supported_ci = SupportedCI.from_ci_env()
    try:
        fcn = PREVIOUS_COMMIT_SHA_FUNCTIONS[supported_ci]
    except KeyError:
        raise UnexpectedError(f"Not implemented for {supported_ci.value}")
    return fcn(verbose)
```

Same thing in commit_range.py.

Reply:
Modified this way, it's indeed consiser